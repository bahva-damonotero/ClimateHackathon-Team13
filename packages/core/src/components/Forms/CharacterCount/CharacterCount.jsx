import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import characterCount from '@uswds/uswds/js/usa-character-count';

export default function CharacterCount({ children, target }) {
  const containerRef = useRef();

  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // add the .usa-character-count__field class to the target input
    const input = containerEl.querySelector(`#${target}`);
    input.classList.add('usa-character-count__field');

    // call USWDS's `on` function to initialize the component
    characterCount.on(containerEl);

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      characterCount.off(containerEl);
    };
  }, []);

  return (
    <div className="bah-uswds-character-count" ref={containerRef}>
      <div className="usa-character-count">
        {children}

        <span
          id={`character-count--${target}`}
          className="usa-hint usa-character-count__message"
          aria-live="polite"
        />
      </div>
    </div>
  );
}

CharacterCount.propTypes = {
  children: PropTypes.node.isRequired,
  /** ID of the input to show the character count for */
  target: PropTypes.string.isRequired
};
