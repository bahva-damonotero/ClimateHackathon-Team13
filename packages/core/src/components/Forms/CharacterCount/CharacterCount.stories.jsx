import React from 'react';

import { Form, FormGroup, Input } from '..';
import CharacterCount from '.';

const idText = 'character-count-example-1';
const idTextArea = 'character-count-example-2';
const maxLength = 25;

export default {
  title: 'Forms/CharacterCount',
  component: CharacterCount,
  parameters: {
    controls: { hideNoControlsWarning: true }
  }
};

export const inputAndTextarea = () => (
  <Form>
    <CharacterCount target={idText}>
      <FormGroup>
        <Input
          describedBy={`character-count--${idText}`}
          id={idText}
          maxLength={maxLength}
        >
          Text input
        </Input>
      </FormGroup>
    </CharacterCount>

    <CharacterCount target={idTextArea}>
      <FormGroup>
        <Input
          describedBy={`character-count--${idTextArea}`}
          id={idTextArea}
          maxLength={maxLength}
          type="textarea"
        >
          Textarea input
        </Input>
      </FormGroup>
    </CharacterCount>
  </Form>
);
