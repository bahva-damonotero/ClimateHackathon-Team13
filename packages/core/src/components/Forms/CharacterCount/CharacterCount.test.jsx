import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import characterCount from '@uswds/uswds/js/usa-character-count';

import CharacterCount from '.';
import { FormGroup, Input } from '..';

describe('CharacterCount', () => {
  const id = 'character-count-test';
  const inputId = 'character-count-test-1';
  const maxLength = 25;

  const jsx = (
    <CharacterCount id={id} target={inputId}>
      <FormGroup>
        <Input
          describedBy={`character-count--${inputId}`}
          id={inputId}
          maxLength={maxLength}
        >
          Text input
        </Input>
      </FormGroup>
    </CharacterCount>
  );

  it('should call `characterCount.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(characterCount, 'on');

    render(jsx);

    expect(onSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
