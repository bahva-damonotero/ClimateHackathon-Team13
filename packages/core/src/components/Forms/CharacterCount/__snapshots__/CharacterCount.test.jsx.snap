// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCount should match snapshot 1`] = `
<div
  className="bah-uswds-character-count"
>
  <div
    className="usa-character-count"
  >
    <div
      className="usa-form-group"
    >
      <label
        className="usa-label"
        htmlFor="character-count-test-1"
      >
        Text input
      </label>
      <input
        aria-describedby="character-count--character-count-test-1"
        className="usa-input"
        id="character-count-test-1"
        maxLength={25}
        name="character-count-test-1"
        type="text"
      />
    </div>
    <span
      aria-live="polite"
      className="usa-hint usa-character-count__message"
      id="character-count--character-count-test-1"
    />
  </div>
</div>
`;
